File: dmitri/publications/graphical-chemistry-generator/src/meibutsu.py
"""Meibutsu: an inverse instrument for chemical structure.

A conventional instrument maps an analyte to a measured value.  This
module implements the opposite direction: it accepts a measured value --
a vibrational spectrum, a coordinate triple, or an address -- and returns
the structure that would produce it.

The implementation is deliberately small and has no dependencies beyond
numpy.  Every quantity it computes is either read from a public reference
database or derived here in closed form; nothing is fitted.

Three layers, in dependency order:

  1. ADDRESSING.  A spectrum maps to a point in the unit cube and thence
     to a base-3 address.  This layer is taken from the reference
     implementation of the compound database and is not re-derived.

  2. OBSERVATION.  An address is realised as a complex field on a
     discrete grid -- an amplitude and a phase at each grid point.  This
     is the quantity a fragment shader would evaluate per pixel.

  3. COMPARISON.  Two observation fields are compared by superposition.
     The comparison is a normalised correlation of the two fields, which
     is the fringe visibility of their superposition.

The third layer is where the design differs from a conventional
pipeline: no feature is extracted from either field, and no similarity
function is evaluated on a pair of feature vectors.  The fields are
added, and the visibility of the resulting fringes is read off.
"""

from __future__ import annotations

import math
import os
import sys
from dataclasses import dataclass, field as _field

import numpy as np

_DB = os.path.join(os.path.dirname(__file__), "..", "..", "..",
                   "categorical-compound-database")
sys.path.insert(0, os.path.abspath(_DB))

from categorical_compound_database import (  # noqa: E402
    COMPOUNDS, CategoricalEncoder, coords_to_trits,
)

# --------------------------------------------------------------------
# Reference scales.  These are properties of the reference database, not
# free parameters: OMEGA_REF is the highest fundamental it contains.
# --------------------------------------------------------------------

OMEGA_REF = 4401.0          # cm^-1, H2 stretch
GRID = 256                  # observation grid points per axis
DEPTH = 18                  # address length in trits

_ENC = CategoricalEncoder(COMPOUNDS)


# ====================================================================
#  Layer 1: addressing
# ====================================================================

def coordinates(modes, b_rot=None):
    """Map a spectrum to its coordinate triple in the unit cube."""
    return (_ENC.compute_S_k(list(modes), b_rot),
            _ENC.compute_S_t(list(modes), b_rot),
            _ENC.compute_S_e(list(modes)))


def address(coords, depth=DEPTH):
    """Map a coordinate triple to its base-3 address."""
    sk, st, se = coords
    return "".join(str(t) for t in coords_to_trits(sk, st, se, depth=depth))


def common_prefix(a: str, b: str) -> int:
    """Length of the longest shared address prefix."""
    n = 0
    for x, y in zip(a, b):
        if x != y:
            break
        n += 1
    return n


# ====================================================================
#  Layer 2: observation
# ====================================================================

@dataclass
class Observation:
    """A complex field on a grid: what one evaluation pass produces.

    ``amp`` and ``phase`` are what a shader writes to two channels of a
    texture; ``field`` is their complex combination.  Storing all three
    is redundant by design -- the redundancy is what lets the comparison
    layer work on the field while the diagnostics work on amplitude and
    phase separately.
    """
    name: str
    coords: tuple
    modes: list
    amp: np.ndarray
    phase: np.ndarray

    @property
    def field(self) -> np.ndarray:
        return self.amp * np.exp(1j * self.phase)

    @property
    def energy(self) -> float:
        return float(np.vdot(self.field, self.field).real)


def observe(modes, coords=None, b_rot=None, grid=GRID, name="") -> Observation:
    """Evaluate the observation field for one spectrum.

    Amplitude places one lobe per vibrational mode at its normalised
    frequency address, with a width set by the first coordinate and an
    envelope set by the second.  Phase accumulates linearly with the
    mode frequencies and with the third coordinate.

    The phase is the load-bearing part.  Amplitude alone would make two
    molecules with the same mode positions indistinguishable regardless
    of their coordinates; the phase term is what carries the coordinate
    information into the comparison.
    """
    modes = list(modes)
    if coords is None:
        coords = coordinates(modes, b_rot)
    sk, st, se = coords

    u = np.linspace(0.0, 1.0, grid)
    width = 0.05 * (1.0 - 0.5 * sk) + 1e-6

    amp = np.zeros(grid)
    for om in modes:
        amp += np.exp(-((u - om / OMEGA_REF) ** 2) / (width ** 2))
    amp *= np.exp(-((u - 0.5) ** 2) / ((0.1 + 0.4 * st) ** 2))

    phase = 2.0 * math.pi * se * u * 8.0
    for om in modes:
        phase = phase + 2.0 * math.pi * (om / OMEGA_REF) * u

    return Observation(name=name, coords=tuple(coords), modes=modes,
                       amp=amp, phase=phase)


# ====================================================================
#  Layer 3: comparison by superposition
# ====================================================================

def superpose(a: Observation, b: Observation) -> np.ndarray:
    """The superposed intensity |A + B|^2.

    This is the only place two observations meet.  Note that it is an
    addition, not a comparison: the two fields coexist on one grid and
    the interference is what results.
    """
    s = a.field + b.field
    return np.abs(s) ** 2


def visibility(a: Observation, b: Observation) -> float:
    """Fringe visibility of the superposition of two observations.

    Defined as the normalised field correlation

        V = |<A, B>| / sqrt(<A,A> <B,B>)

    which is exactly 1 when the two fields are identical (Cauchy-Schwarz
    with equality iff the fields are proportional) and 0 when they are
    orthogonal.  No assumption about the phase distribution is required
    for the self-comparison case, which is why this form is used in
    preference to a pointwise expression averaged over the grid.

    The cross-term of the superposition carries the entire relational
    content: |A+B|^2 = |A|^2 + |B|^2 + 2 Re<A,B>.  The first two terms
    are properties of each observation alone.
    """
    fa, fb = a.field, b.field
    den = math.sqrt(a.energy * b.energy)
    if den <= 0.0:
        return 0.0
    return float(np.abs(np.vdot(fa, fb)) / den)


def cross_term(a: Observation, b: Observation) -> np.ndarray:
    """The relational part of the superposition, pointwise.

    Everything that depends on BOTH observations lives here; the
    remaining terms of |A+B|^2 would be present if the other observation
    were absent.
    """
    return 2.0 * np.real(np.conj(a.field) * b.field)


def stack(observations) -> np.ndarray:
    """Superpose an arbitrary number of observations at once.

    Superposition is linear, so this is a single addition regardless of
    how many observations are supplied -- no pairwise loop appears.  The
    resulting intensity contains every pairwise cross-term.
    """
    total = np.zeros_like(observations[0].field)
    for o in observations:
        total = total + o.field
    return np.abs(total) ** 2


def stacked_cross_energy(observations) -> float:
    """Total relational content of a stack.

    |sum A_i|^2 summed over the grid, minus the sum of the individual
    energies, leaves exactly the sum of all pairwise cross-terms.  This
    identity is what makes bulk comparison possible: the quantity is
    obtained from one superposition, not from C(n,2) comparisons.
    """
    tot = float(np.sum(stack(observations)))
    own = sum(o.energy for o in observations)
    return tot - own


# ====================================================================
#  The inverse direction
# ====================================================================

@dataclass
class Resolution:
    """What the instrument returns when given a measured value.

    ``occupants`` may be empty (the address names an unoccupied cell),
    a single structure, or several (the cell is not resolved at this
    depth).  All three are reported rather than collapsed, because the
    distinction between them is the instrument's own statement about
    the resolution it achieved.
    """
    query_coords: tuple
    query_address: str
    depth: int
    occupants: list = _field(default_factory=list)
    fallback_depth: int = 0
    ranked: list = _field(default_factory=list)

    @property
    def resolved(self) -> bool:
        return len(self.occupants) == 1


class Instrument:
    """The inverse instrument over a reference set of structures."""

    def __init__(self, compounds=None, grid=GRID, depth=DEPTH):
        self.db = compounds if compounds is not None else COMPOUNDS
        self.grid = grid
        self.depth = depth
        self.records = {}
        for name in self.db:
            rec = _ENC.encode(name)
            self.records[name] = rec
        # Coordinates are recomputed rather than read back from the
        # stored record: the encoder rounds to six decimals when it
        # stores, and feeding a rounded coordinate into the field
        # perturbs every downstream visibility at the 1e-6 level for no
        # modelling reason.  The rounding is a property of the storage
        # format, not of the construction.
        self.obs = {
            n: observe(r["modes"],
                       coordinates(r["modes"], self.db[n].get("B_rot")),
                       grid=grid, name=n)
            for n, r in self.records.items()
        }
        self.addresses = {n: r["trit_string"] for n, r in self.records.items()}

    # -- inverse: measured value in, structure out --------------------

    def resolve(self, modes, b_rot=None, depth=None, rank=True) -> Resolution:
        """Given a spectrum, return the structures that would produce it."""
        depth = depth or self.depth
        coords = coordinates(modes, b_rot)
        addr = address(coords, depth)

        occupants = [n for n, a in self.addresses.items()
                     if a[:depth] == addr[:depth]]

        # fall back to the deepest occupied prefix
        fb = depth
        while not occupants and fb > 0:
            fb -= 1
            occupants = [n for n, a in self.addresses.items()
                         if a[:fb] == addr[:fb]]

        ranked = []
        if rank:
            q = observe(modes, coords, grid=self.grid, name="<query>")
            ranked = sorted(
                ((n, visibility(q, self.obs[n])) for n in self.db),
                key=lambda kv: -kv[1])
        return Resolution(query_coords=coords, query_address=addr,
                          depth=depth, occupants=sorted(occupants),
                          fallback_depth=fb, ranked=ranked)

    # -- comparison ---------------------------------------------------

    def visibility_matrix(self):
        names = sorted(self.obs)
        n = len(names)
        M = np.zeros((n, n))
        for i, a in enumerate(names):
            for j, b in enumerate(names):
                M[i, j] = visibility(self.obs[a], self.obs[b])
        return names, M

    def bulk(self, names=None):
        """One superposition over many structures."""
        names = names or sorted(self.obs)
        return stack([self.obs[n] for n in names])


__all__ = [
    "OMEGA_REF", "GRID", "DEPTH", "coordinates", "address", "common_prefix",
    "Observation", "observe", "superpose", "visibility", "cross_term",
    "stack", "stacked_cross_energy", "Resolution", "Instrument",
]

File: honjo-masamune/src/lib/meibutsu.js
/**
 * Observation fields and comparison by superposition.
 *
 * This is a port of the Python reference implementation
 * (graphical-chemistry-generator/src/meibutsu.py). It is checked
 * against it by scripts/check-field.mjs, which compares coordinates,
 * visibilities and energies to 1e-9. If the two drift apart that
 * script fails.
 *
 * The reason this exists in JavaScript at all: the comparison is a
 * superposition of two fields, and the superposition is what gets
 * drawn. Computing it on a server and shipping a picture of the result
 * would make the displayed thing a report about a comparison rather
 * than the comparison itself.
 */

/** Highest fundamental in the reference set: H2 stretch, cm^-1. */
export const OMEGA_REF = 4401.0;

/** Reference scales for the temporal coordinate. */
const OMEGA_REF_MAX = 4401.0;
const OMEGA_REF_MIN = 218.0;
const B_ROT_REF_MIN = 0.39;

/** Rational-ratio tolerance for the harmonic coordinate. */
const HARMONIC_DELTA = 0.05;

/* ------------------------------------------------------------------ */
/*  Coordinates                                                       */
/* ------------------------------------------------------------------ */

/**
 * Knowledge coordinate: normalised Shannon entropy of the frequency
 * distribution. One dominant mode gives 0; equal modes give 1.
 */
export function knowledge(modes) {
  const n = modes.length;
  if (n === 0) return 0;
  if (n === 1) return modes[0] / OMEGA_REF_MAX;
  const total = modes.reduce((a, b) => a + b, 0);
  let h = 0;
  for (const w of modes) {
    const p = w / total;
    if (p > 0) h -= p * Math.log2(p);
  }
  return Math.min(h / Math.log2(n), 1.0);
}

/**
 * Temporal coordinate: how many decades of timescale the spectrum
 * spans, against the reference span.
 */
export function temporal(modes, bRot) {
  const n = modes.length;
  if (n === 1) {
    if (bRot && bRot > 0) {
      const ratio = modes[0] / bRot;
      if (ratio <= 1) return 0;
      const v = Math.log(ratio) / Math.log(OMEGA_REF_MAX / B_ROT_REF_MIN);
      return Math.min(Math.max(v, 0), 1);
    }
    return 0.1;
  }
  const hi = Math.max(...modes);
  const lo = Math.min(...modes);
  if (lo <= 0 || hi <= lo) return 0;
  const v = Math.log(hi / lo) / Math.log(OMEGA_REF_MAX / OMEGA_REF_MIN);
  return Math.min(Math.max(v, 0), 1);
}

function isHarmonicPair(a, b, delta = HARMONIC_DELTA) {
  if (a === 0 || b === 0) return false;
  const ratio = Math.max(a, b) / Math.min(a, b);
  for (let p = 1; p <= 8; p += 1) {
    for (let q = 1; q <= p; q += 1) {
      if (Math.abs(ratio - p / q) < delta) return true;
    }
  }
  return false;
}

/**
 * Evolution coordinate: the fraction of mode pairs whose frequency
 * ratio lands on a low-order rational. This is the molecule
 * interfering with itself.
 */
export function evolution(modes) {
  const n = modes.length;
  if (n < 2) return 0;
  const pairs = (n * (n - 1)) / 2;
  let harmonic = 0;
  for (let i = 0; i < n; i += 1) {
    for (let j = i + 1; j < n; j += 1) {
      if (isHarmonicPair(modes[i], modes[j])) harmonic += 1;
    }
  }
  return harmonic / Math.max(pairs, 1);
}

export function coordinates(modes, bRot) {
  return [knowledge(modes), temporal(modes, bRot), evolution(modes)];
}

/* ------------------------------------------------------------------ */
/*  The observation field                                             */
/* ------------------------------------------------------------------ */

/**
 * Evaluate the field for one spectrum.
 *
 * Amplitude: one Gaussian lobe per mode at its normalised frequency
 * address, width set by the knowledge coordinate, under an envelope set
 * by the temporal coordinate.
 *
 * Phase: accumulates linearly with the modes and with the evolution
 * coordinate. This is the load-bearing part — amplitude alone cannot
 * distinguish structures whose mode positions coincide.
 */
export function observe(modes, { coords = null, bRot = null, grid = 256, name = "" } = {}) {
  const c = coords || coordinates(modes, bRot);
  const [sk, st, se] = c;
  const width = 0.05 * (1.0 - 0.5 * sk) + 1e-6;
  const env = 0.1 + 0.4 * st;

  const amp = new Float64Array(grid);
  const phase = new Float64Array(grid);

  for (let i = 0; i < grid; i += 1) {
    const u = grid === 1 ? 0 : i / (grid - 1);
    let a = 0;
    let ph = 2 * Math.PI * se * u * 8.0;
    for (const om of modes) {
      const mu = om / OMEGA_REF;
      const d = u - mu;
      a += Math.exp(-(d * d) / (width * width));
      ph += 2 * Math.PI * mu * u;
    }
    const e = u - 0.5;
    amp[i] = a * Math.exp(-(e * e) / (env * env));
    phase[i] = ph;
  }

  return { name, coords: c, modes: [...modes], amp, phase, grid };
}

/** Squared norm of the field. */
export function energy(f) {
  let s = 0;
  for (let i = 0; i < f.grid; i += 1) s += f.amp[i] * f.amp[i];
  return s;
}

/**
 * Fringe visibility: the normalised field correlation
 *
 *     V = |<A,B>| / sqrt(<A,A> <B,B>)
 *
 * Exactly 1 when the fields coincide, by Cauchy-Schwarz with equality.
 * This is used in preference to a pointwise average of
 * 1/2 + 1/2 cos(dphi), which attains only ~0.53 on self-comparison and
 * requires an unstated assumption about the phase distribution.
 */
export function visibility(a, b) {
  let re = 0;
  let im = 0;
  let ea = 0;
  let eb = 0;
  for (let i = 0; i < a.grid; i += 1) {
    const d = b.phase[i] - a.phase[i];
    const m = a.amp[i] * b.amp[i];
    re += m * Math.cos(d);
    im += m * Math.sin(d);
    ea += a.amp[i] * a.amp[i];
    eb += b.amp[i] * b.amp[i];
  }
  const den = Math.sqrt(ea * eb);
  if (den <= 0) return 0;
  return Math.hypot(re, im) / den;
}

/**
 * The relational part of the superposition, pointwise.
 *
 * |A+B|^2 = |A|^2 + |B|^2 + 2 Re(conj(A) B)
 *
 * Everything depending on both fields is this third term; the other
 * two would be present if the second structure were absent.
 */
export function crossTerm(a, b) {
  const out = new Float64Array(a.grid);
  for (let i = 0; i < a.grid; i += 1) {
    out[i] = 2 * a.amp[i] * b.amp[i] * Math.cos(b.phase[i] - a.phase[i]);
  }
  return out;
}

/** Superposed intensity |A + B|^2. */
export function superpose(a, b) {
  const out = new Float64Array(a.grid);
  for (let i = 0; i < a.grid; i += 1) {
    const d = b.phase[i] - a.phase[i];
    out[i] =
      a.amp[i] * a.amp[i] +
      b.amp[i] * b.amp[i] +
      2 * a.amp[i] * b.amp[i] * Math.cos(d);
  }
  return out;
}

/**
 * Render a field as a 2-D interference pattern.
 *
 * The horizontal axis is the frequency address; the vertical axis is a
 * phase offset applied to the second field. A horizontal slice at
 * offset 0 is the superposition as computed above; the other rows show
 * how the pattern moves as the relative phase is swept, which is what
 * makes the fringes visible as fringes rather than as a single curve.
 *
 * Writes straight into an RGBA buffer: this array is both what is
 * displayed and what the visibility is computed from.
 */
export function renderInterference(a, b, width, height) {
  const buf = new Uint8ClampedArray(width * height * 4);
  const rows = height;

  // Normalise against the largest intensity anywhere in the sweep, so
  // brightness is comparable between molecule pairs.
  let peak = 0;
  for (let r = 0; r < rows; r += 1) {
    const off = (r / Math.max(rows - 1, 1)) * 2 * Math.PI;
    for (let i = 0; i < a.grid; i += 1) {
      const d = b.phase[i] - a.phase[i] + off;
      const v =
        a.amp[i] * a.amp[i] +
        b.amp[i] * b.amp[i] +
        2 * a.amp[i] * b.amp[i] * Math.cos(d);
      if (v > peak) peak = v;
    }
  }
  if (peak <= 0) peak = 1;

  for (let r = 0; r < rows; r += 1) {
    const off = (r / Math.max(rows - 1, 1)) * 2 * Math.PI;
    for (let x = 0; x < width; x += 1) {
      // sample the field grid across the canvas width
      const gi = Math.min(
        a.grid - 1,
        Math.round((x / Math.max(width - 1, 1)) * (a.grid - 1))
      );
      const d = b.phase[gi] - a.phase[gi] + off;
      const own = a.amp[gi] * a.amp[gi] + b.amp[gi] * b.amp[gi];
      const cross = 2 * a.amp[gi] * b.amp[gi] * Math.cos(d);
      const t = Math.max(0, Math.min(1, (own + cross) / peak));

      // Colour carries the sign of the cross-term: constructive
      // interference warm, destructive cool. A single-channel ramp
      // would hide which of the two is happening.
      const k = (r * width + x) * 4;
      const sign = cross >= 0 ? 1 : -1;
      const mag = Math.sqrt(t);
      if (sign > 0) {
        buf[k] = 40 + 215 * mag;
        buf[k + 1] = 40 + 160 * mag;
        buf[k + 2] = 60 + 60 * mag;
      } else {
        buf[k] = 30 + 60 * mag;
        buf[k + 1] = 60 + 140 * mag;
        buf[k + 2] = 80 + 215 * mag;
      }
      buf[k + 3] = 255;
    }
  }
  return buf;
}

/** Draw a single field's amplitude and phase as a strip. */
export function renderField(f, width, height) {
  const buf = new Uint8ClampedArray(width * height * 4);
  let peak = 0;
  for (let i = 0; i < f.grid; i += 1) if (f.amp[i] > peak) peak = f.amp[i];
  if (peak <= 0) peak = 1;

  for (let x = 0; x < width; x += 1) {
    const gi = Math.min(
      f.grid - 1,
      Math.round((x / Math.max(width - 1, 1)) * (f.grid - 1))
    );
    const a = f.amp[gi] / peak;
    const ph = ((f.phase[gi] % (2 * Math.PI)) + 2 * Math.PI) % (2 * Math.PI);
    const hue = ph / (2 * Math.PI);
    const [r, g, bl] = hsv(hue, 0.55, 0.25 + 0.75 * a);
    for (let y = 0; y < height; y += 1) {
      // amplitude sets how far up the column is lit
      const lit = y / height >= 1 - a;
      const k = (y * width + x) * 4;
      buf[k] = lit ? r : 26;
      buf[k + 1] = lit ? g : 27;
      buf[k + 2] = lit ? bl : 38;
      buf[k + 3] = 255;
    }
  }
  return buf;
}

function hsv(h, s, v) {
  const i = Math.floor(h * 6);
  const f = h * 6 - i;
  const p = v * (1 - s);
  const q = v * (1 - f * s);
  const t = v * (1 - (1 - f) * s);
  let r;
  let g;
  let b;
  switch (i % 6) {
    case 0: [r, g, b] = [v, t, p]; break;
    case 1: [r, g, b] = [q, v, p]; break;
    case 2: [r, g, b] = [p, v, t]; break;
    case 3: [r, g, b] = [p, q, v]; break;
    case 4: [r, g, b] = [t, p, v]; break;
    default: [r, g, b] = [v, p, q];
  }
  return [Math.round(r * 255), Math.round(g * 255), Math.round(b * 255)];
}
